File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/IndoorBuildingFunc.java
package com.sdoward.rxgooglemap;

import com.google.android.gms.maps.*;
import com.google.android.gms.maps.model.*;
import com.sdoward.rxgooglemap.events.*;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.functions.Function;

class IndoorBuildingFunc implements Function<GoogleMap, Observable<IndoorBuildingEvent>> {

    @Override
    public Observable<IndoorBuildingEvent> apply(final GoogleMap googleMap) {
        return Observable.create(new ObservableOnSubscribe<IndoorBuildingEvent>() {
            @Override
            public void subscribe(final ObservableEmitter<IndoorBuildingEvent> emitter) {
                googleMap.setOnIndoorStateChangeListener(
                        new GoogleMap.OnIndoorStateChangeListener() {
                            @Override
                            public void onIndoorBuildingFocused() {
                                emitter.onNext(new IndoorBuildingEvent());
                            }

                            @Override
                            public void onIndoorLevelActivated(IndoorBuilding indoorBuilding) {
                                emitter.onNext(new IndoorLevelActivatedEvent(indoorBuilding));
                            }
                        });
            }
        });
    }
}

File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/IndoorBuildingFunc.kt
package com.sdoward.rxgooglemap

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.IndoorBuilding
import com.sdoward.rxgooglemap.events.IndoorBuildingEvent
import com.sdoward.rxgooglemap.events.IndoorLevelActivatedEvent
import io.reactivex.Observable
import io.reactivex.Observable.create
import io.reactivex.functions.Function


class IndoorBuildingFunc : Function<GoogleMap, Observable<IndoorBuildingEvent>> {

    override fun apply(t: GoogleMap): Observable<IndoorBuildingEvent> {
        return create { subscriber ->
            t.setOnIndoorStateChangeListener(object : GoogleMap.OnIndoorStateChangeListener {
                override fun onIndoorBuildingFocused() {
                    subscriber.onNext(IndoorBuildingEvent())
                }

                override fun onIndoorLevelActivated(p0: IndoorBuilding?) {
                    subscriber.onNext(p0?.let { IndoorLevelActivatedEvent(it) })

                }

            })
        }
    }

}